### PR TITLE
[SPARK-38951][SQL] Aggregate aliases override field names in ResolveAggregateFunctions

### PR DESCRIPTION
**Pull request by cxzl25**
### What changes were proposed in this pull request?
When the expression of `sortOrder` has the name of `UnresolvedAttribute` in the alias of `Aggregate`, use the original expression.

### Why are the changes needed?

The alias of `Aggregate` overwrites the field name, which causes an error in the `sortOrder` field parsing, and then the SQL cannot be executed.

```sql
select sum(id) as id  from range(10) group by id order by sum(id);
```

```
Error in query: Resolved attribute(s) id#0L missing from id#1L in operator !Aggregate [id#1L], [sum(id#1L) AS id#0L, sum(id#0L) AS sum(id#0L)#4L]. Attribute(s) with the same name appear in the operation: id. Please check if the right attribute(s) are used.;
Project [id#0L]
+- Sort [sum(id#0L)#4L ASC NULLS FIRST], true
   +- !Aggregate [id#1L], [sum(id#1L) AS id#0L, sum(id#0L) AS sum(id#0L)#4L]
      +- Range (0, 10, step=1, splits=None)
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
add UT

### CODE REVIEW

**Comment by cxzl25** [comment]
Help review the problem that alias overwrites the original field name, resulting in order by parsing failure.
@cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 1]
```diff
@@ -1176,4 +1176,13 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
     }
   }
+
+  test("SPARK-38951: Aggregate aliases override field names in ResolveAggregateFunctions") {
+    assertAnalysisSuccess(parsePlan(
+      s"""
+         |select sum(id) as id
+         |from range(10)
+         |group by id
+         |order by sum(id)""".stripMargin))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:1186]: does this query work on other databases?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -2563,7 +2563,23 @@ class Analyzer(override val catalogManager: CatalogManager)
 
       case Sort(sortOrder, global, agg: Aggregate) if agg.resolved =>
         // We should resolve the references normally based on child (agg.output) first.
-        val maybeResolved = sortOrder.map(_.child).map(resolveExpressionByPlanOutput(_, agg))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2566]: I think this bug is much more complicated. Think about `order by sum(id)` and `order by abs(id)`. For the first case, we want to resolve `id` to the table column and push `sum(id)` to the Aggregate. For the second case, we want to resolve `id` to `sum(id) as id`.

How to define a clear rule here?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2566]: My proposal is: we still try to resolve the column to `agg.output` first, but temporarily (use `TempResolvedColumn`). If the order by expression is resolved but contains aggregate functions, unresolve the column and call `resolveOperatorWithAggregate`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2566]: e.g.
```
val maybeResolved = sortOrder.map(_.child).map { expr =>
  val resolved = resolveTemp(expr, agg)
  if (resolved.exists(_.isInstanceOf[AggregateFunction])) {
    unresolve(resolved)
  } else {
    resolved
  }
}
resolveOperatorWithAggregate...
```

**Comment by cloud-fan** [comment]
Have you run this query with other databases like postgres? I think the SQL semantic is `order by` col should resolve to columns in SELECT list first, not the table columns. Or I misunderstand it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -2594,6 +2601,31 @@ class Analyzer(override val catalogManager: CatalogManager)
         })
     }
 
+    private def resolveTemp(expr: Expression, agg: Aggregate): Expression = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2604]: This is very similar to the `resolveCol` method inside `resolveExprsWithAggregate`. Can we have a common method to share code?

**Comment by smallzhongfeng** [comment]
Not only in the order by statement, but also when filtering with having, you can see [https://issues.apache.org/jira/browse/SPARK-41236](https://issues.apache.org/jira/browse/SPARK-41236) @cloud-fan @cxzl25

**Pull request closed without merging**